Initialize platform in trng test

RonEld opened the pull request:
### Description

<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->
Add calls to `mbedtls_platform_setup()` and
`mbedtls_platform_terminate()` to the trng greentea test, to
initialize the hardware acceleration engines, in some platforms.
However, this involves hal test including Mbed TLS API, which I am not sure is architecturally correct. On the other hand, `mbedtls_platform_setup()` is intended for initializing the HW acceleration engines, where needed.

This should fix IOTPART-6810

@ashok-rao Could you please check that this resolves the issue you encountered on LAIRD BL654, and if it does, please return the usage of Cryptocell in that library?

### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [x] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

<!-- 
    Optional
    Request additional reviewers with @username
-->

@ARMmbed/mbed-os-crypto

Commit:
Initialize platform in trng test

Add calls to `mbedtls_platform_setup()` and
`mbedtls_platform_terminate()` to the trng greentea test, to
initialize the hardware acceleration engines, in some platforms.

ciarmcom commented:
@RonEld, thank you for your changes.
@ARMmbed/mbed-os-hal @ARMmbed/mbed-os-test @ARMmbed/mbed-os-maintainers please review.

dannybenor commented:
@Patater are there any other initialization functions required? @bulislaw  Should this and other init functions called at some global part of mbedos?

RonEld commented:
Note:
- `psa_crypto_init()` is already called inside `inject_entropy_for_psa()`, in line 85, so no other psa crypto initialization function required.

dannybenor commented:
@RonEld this function should not be called in the test, but in mbed OS. We need the test to make sure mbed OS will work. Please move the call to a platform specific initialization code

dannybenor reviewed (changes requested):
The call to mbedtls_platform_setup should be part of mbed OS and not the test

RonEld commented:
@dannybenor Although I agree with you, this is the design of the function, and of Mbed OS design.
It was original in the Mbed OS boot, but after long discussions, it was decided it should be called by the applications requesting Mbed TLS \ crypto operations.
@bulislaw please confirm that this is still the case

RonEld commented:
More information in IOTSSL-1576

dannybenor commented:
@bulislaw need your input here. We have functionality in mbedOS that uses crypto. Adding a new target should not brake it. Adding the initialization to the test is not helping. If the function mbedtls_platform_terminate() is required for some targets, code should not work for any target if the function is not called.

RonEld commented:
In current design, modules in Mbed OS, that use Mbed TLS \ crypto, should call `mbedtls_platform_setup()` and `mbedtls_platform_terminate()`, like BLE does.
We have added a reference counter in the platform context, that will not actually terminate the platform hw acceleration, if still used. 
See https://github.com/ARMmbed/mbed-os/pull/7099

dannybenor commented:
@liatvindzberg @erezlandau FYI
@RonEld @sbutcher-arm why the function mbedtls_platform_setup() is not called inside psa_crypto_init()?
The implementation of devicekey and securestore passed a lot of review of the crypto team and nobody pointed that we need to call the function mbedtls_platform_setup(). That means we are setting a big trap for any developer.
The parameter mbedtls_platform_context looks like something not required if no other TLS function uses it.
https://context.reverso.net/translation/hebrew-english/%D7%91%D7%90%D7%A8%D7%96%D7%99%D7%9D+%D7%A0%D7%A4%D7%9C%D7%94+%D7%A9%D7%9C%D7%94%D7%91%D7%AA

RonEld commented:
>why the function mbedtls_platform_setup() is not called inside psa_crypto_init()?

This is in discussion in https://github.com/ARMmbed/mbed-crypto/issues/24 . IMHO, once PSA will be fully integrated within Mbed OS, `mbedtls_platfor_setup()` will be deprecated, or at least not handle the crypto operations. This function precedes PSA crypto.

>The implementation of devicekey and securestore passed a lot of review of the crypto team and nobody pointed that we need to call the function mbedtls_platform_setup().

For PSA point of view, `mbedtls_platform_setup()` should not be called, and as I mentioned in previous answer, it should probably not be used once all crypto operations be called in SA crypto. Until that happens, `mbedtls_platform_setup()` and `mbedtls_platform_terminate()` should be called as well, and I think it's a mistake \ overlook that you weren't been pointed out on that.

bulislaw commented:
That's a breaking change isn't it? As in current use of TLS/crypto won't work without this change? Generally speaking, my personal opinion is, that we shouldn't initialise modules globally unless they are used by the OS itself. In this case user is well aware that he/she will need TLS/crypto so they should initialise it. That being said, it's a breaking change, is there a way for us to hide it in existing calls so that we are backward compatible? Could public API initialise the modules if they weren't initialised before?

RonEld commented:
I wouldn't say it's a breaking change. but a bug fix. 
Without this fix, platforms that implemented their alternative platform initialization function( at the moment only NRF52840 AFAIK) will simply fail on this test. Other platforms will use the default implementation of `mbedtls_platform_setup()` which is empty in Mbed TLS.
Ever since this API was introduced, in `mbedtls-2.7` and then `mbed-os-5.8`, this function was recomended to be used whenever a cryptographic operation is used. See https://os.mbed.com/docs/mbed-os/v5.11/porting/hardware-accelerated-crypto.html#mbed-tls-platform-context and IOTSSL-1576

>That being said, it's a breaking change, is there a way for us to hide it in existing calls so that we are backward compatible? Could public API initialise the modules if they weren't initialised before?

As mentioned, I don't think it's a breaking change, but a bug fix, and we are backwards compatible. The NRF52480 didn't work before on this test, since the Cryptocell library porting was introduced in #6794
and now it does( It was a breaking change in that PR though). As for initialize the modules if they weren't initialized before, well, since #7099, there is a refernce counter in the platform context, so it is posisble, however, this requires *all* the Mbed TLS APIto call this function, which means changing Mbed TLS for this transition to mbed crypto stage, which I think is redundant.  note that in https://github.com/ARMmbed/mbed-crypto/issues/24 we are discussing how to integrate this function with `psa_crypto_init()`, but until this is done, and psa crypto is fully incorporated in Mbed OS, I think this change is needed.

0xc0170 commented:
This "simple" fix seems to be difficult to understand the implications. @RonEld Where is this captured (any docs) that these 2 functions need to be invoked when crypto used ? How we can proceed further?

@ARMmbed/mbed-os-crypto Please review

RonEld commented:
@0xc0170 As mentioned in previous comment, it is described in https://os.mbed.com/docs/mbed-os/v5.11/porting/hardware-accelerated-crypto.html#mbed-tls-platform-context

RonEld commented:
Also in the function description: https://github.com/ARMmbed/mbed-os/blob/master/features/mbedtls/inc/mbedtls/platform.h#L335

yanesca reviewed (approved):
The implications of this fix:
- The test will pass for every properly functioning and integrated TRNG
- The test won't catch that some part of the OS uses the TRNG but does not call `mbedtls_platform_setup()` before.
- We can't see yet how this affects future plans (see https://github.com/ARMmbed/mbed-crypto/issues/24)

If we are unhappy with the fix we have two other options:
1. Leave it as it is and have correct implementations fail the test suite.
2. Move `mbedtls_platform_setup()` back to Mbed OS boot.

This fix is a clear improvement to option 1 and I can see why we don't want option 2.

That being said this test does not catch that a part of the OS uses TRNG but does not call `mbedtls_platform_setup()`. This failure can potentially be silent and still catastrophic for security.  Optimally we should identify every part of the OS that uses TRNG and add a test for that module that makes sure that the platform initialisation has been done where it is needed.

I am approving since this fix is a clear improvement, but it would be best to cover the gaps in the test suite. If that is out of scope for this PR, then we should raise an issue about it. (My approval is conditional on fixing the gap or raising an issue for it.)

0xc0170 commented:
Thanks @yanesca ! @dannybenor Do you agree with the above statement about options and approval for this PR?

dannybenor commented:
O do not agree with adding the call to the test.
We need to find a better way to integrate into mbed os

RonEld commented:
From what you are saying, you also do not agree of having `psa_crypto_init` in the test, as it is basically intended to do the same thing, once crypto will be fully integrated

RonEld commented:
Note: Having the initialization done at boot time, causing the hw acceleration engine to always be "on", might have side effects of high power consumption. see https://github.com/ARMmbed/mbed-os/pull/9372 for example.

simonbutcher commented:
> @RonEld @sbutcher-arm why the function mbedtls_platform_setup() is not called inside psa_crypto_init()?
> The implementation of devicekey and securestore passed a lot of review of the crypto team and nobody pointed that we need to call the function mbedtls_platform_setup(). That means we are setting a big trap for any developer.

At the moment we're moving the crypto APIs from Mbed TLS to PSA. We haven't completed that work yet, and it's still in transition. When we have completed that work, I'd expect users of *only* Crypto APIs to not need to call `mbedtls_platform_setup()`. That's not true yet, but it should become true when we've finished the work.

cc: @Patater

0xc0170 commented:
@RonEld Can you talk to @dannybenor to resolve this please? We shall find a path forward or close this PR

dannybenor commented:
@RonEld This PR can be enhanced to make sure all uses of crypto in mbed OS will work, not only this test.
Will  be good also to make sure it is clear in the documentation

RonEld commented:
@dannybenor AFAIK, all uses of Mbed TLS already call this API.
This test was overlooked probably because it is a hal test,

I will work on modifying the documentation.

dannybenor commented:
devicekey, SecureStore ?

RonEld commented:
I will check, perhaps they were missed as well.

kjbracey commented:
This is kind of tangential, but if you're visiting this area, please take more care on the thread safety as you revise it.

I can see there's an atomic counter reference count in `mbedtls_platform_setup` (I was searching for in-tree atomics), but that is not actually safely atomically incremented (counter could reach 2 with no thread attempting setup), and even if it was, there's no threading protection for the setup and teardown:

* Thread B can assume it's already initialised when they increment to 2, even though Thread A is still initialising it.
* Thread B can start initialising it when they increment to 1, even though Thread A is still tearing it down.

I'm assuming the "setup" and "teardown" are not frequent operations - if that's the case, the complexity of atomics isn't justified. Just slap in a `SingletonPtr<PlatformMutex>` for the setup and teardown functions. (And for space efficiency, or even correctness(?), you could share that with mutex with `mbedtls_hardware_poll`, which currently has its own.)

If they are frequent operations, so you want a fast case for 1<->2 reference count transitions, let me know, and I can try to figure out a speed-efficient locking pattern combining atomics and a mutex.

RonEld commented:
@kjbracey-arm 
This PR is not revising `mbedtls_platform_setup()` API, but adding missing calls to it in the code. Nonetheless, if there are issues, I will look into them.
As you can see in [this discussion](https://github.com/ARMmbed/mbed-os/pull/7099#discussion_r213785914) we considered thread safety. I will add a mutex, however I don't think it should be shared with `mbedtls_hardware_poll()`. `mbedtls_hardware_poll()` should *always* be called after a setup has already occurred, so once we reach it, setup has platform setup has already been called, at least once.

I wouldn't say "setup" and "teardown" are frequently called, depending on the caller's implementation, but they could be called from different threads and modules. They should actually be called when an application\feature wants to use Mbed TLS crypto operations, as it was intended to initialize the hardware crypto engine. For example, an application that uses crypto operations will need to call these funtions, but within, BLE is using crypto operations as well, and calls these functions as well. They *could* be called at the same time.


As mentioned in  https://github.com/ARMmbed/mbed-os/pull/9493#issuecomment-462343020 , It will be changed, once the crypto APIs will be transitioned into PSA.

kjbracey commented:
The reason to share the mutex is to save RAM - I agree there's no correctness need. The drawback is unnecessary contention on the shared mutex, but if setup and teardown are not being called frequently, that would be minimal.  (I propose eliminating the atomic reference count altogether and just adjusting it inside the mutex - I don't think this case merits the complexity of an optimised "don't take the lock" path.)

RonEld commented:
very well, I will share the mutex.

@Patater Any thoughts on removing the atomic counter?

Patater commented:
@RonEld I like @kjbracey-arm's suggestions. Removing the atomic counter makes sense, since doing the initialization operation is not atomic despite our attempts at it. Wrapping initialization in a mutex will make initialization atomic and we can do reference counting with the mutex held. Sharing the mutex with `mbedtls_hardware_poll()` should be OK; `mbedtls_hardware_poll()` shouldn't be called until the platform is initialized anyway, initializing things like e.g. the CC310 TRNG.

It's great to have synchronization primitives available in the HAL now. :)

Commit:
Add mbedtls platform setup and teardown to modules

Add calls to `mbedtls_platform_setup()`
and `mbedtls_platform_teardown()` to all modules and tests using Mbed TLS.

Commit:
Use a mutex in platform setup \ teardown functions

Use a singleton Mutex in platforms_alt functions, to be shared
with the trng function, to save RAM. Rename `platform_alt.c`
to `platform_alt.cpp` as the mutex is in a `singletonPtr`
template class.

RonEld commented:
I added calls to `mbedtls_platform_setup()` and `mbedtls_platform_teardown()` to all components and features I saw were using Mbed TLS. I didn't add to PSA using modules, as in PSA targets, there shouldn't be a need to call the Mbed TLS platform functionality for crypto usage.
I rebased, to verify no modules have been missed.

I also added a mutex, and removed the atomic counter, as suggested by @kjbracey-arm

RonEld commented:
>It's great to have synchronization primitives available in the HAL now. :)

Yes, unfortunately, they work for c++ only, as these are classes and templates.

dannybenor commented:
@davidsaada @yossi2le please review

RonEld commented:
@0xc0170 There was a CI failure, so I made a fix for that.
Once CI succeeds, I will squash https://github.com/ARMmbed/mbed-os/pull/9493/commits/cea8790b30b8d162a17eb45fe9e075b5a9b71eb0 with https://github.com/ARMmbed/mbed-os/pull/9493/commits/dcc194fe3c544f8e5426c6e12f8d974045de195d

yossi2le reviewed (approved):
LGTM

cmonr reviewed (approved):
LGTM
